docs/how-to/write-plans.md

```diff
@@ -1,173 +1,93 @@
-
-To enable blueapi to expose the stubs that it makes sense to, but not the others, blueapi will only expose a subset of `MsgGenerator`\ s under the following conditions:
-
-| `__init__.py` in directory has `__exports__`: List[str]: only
-  those named in `__exports__`
-| `__init__.py` in directory has `__all__`: List[str] but no
-  `__exports__`: only those named in `__all__`
-
-This allows other python packages (such as `plans`) to access every function in `__all__`, while only allowing a subset to be called from blueapi as standalone.
-
-``` 
-    # Rehomes all of the beamline's devices. May require to be run standalone
-    from .package import rehome_devices
-    # Awaits a standard callback from analysis. Should not be run standalone
-    from .package import await_callback
-
-    # Exported from the module for use by other modules
-    __all__ = [
-        "rehome_devices",
-        "await_callback",
-    ]
-
-    # Imported by instances of blueapi and allowed to be run
-    __exports__ = [
-        "rehome_devices",
-    ]
+Blueapi exposes the docstrings of plans to clients, along with the parameter types. It is therefore worthwhile to make these detailed and descriptive. This may include units of arguments (e.g. seconds or microseconds), its purpose in the function, the purpose of the plan etc.
+
+```python
+def temp_pressure_snapshot(
+    detectors: List[Readable],
+    temperature: Movable = inject("sample_temperature"),
+    pressure: Movable = inject("sample_pressure"),
+    target_temperature: float = 273.0,
+    target_pressure: float = 10**5,
+    metadata: Optional[Mapping[str, Any]] = None,
+) -> MsgGenerator:
+    """
+    Moves devices for pressure and temperature (defaults fetched from the context)
+    and captures a single frame from a collection of devices
+    Args:
+        detectors (List[Readable]): A list of devices to read while the sample is at STP
+        temperature (Optional[Movable]): A device controlling temperature of the sample,
+            defaults to fetching a device name "sample_temperature" from the context
+        pressure (Optional[Movable]): A device controlling pressure on the sample,
+            defaults to fetching a device name "sample_pressure" from the context
+        target_pressure (Optional[float]): target temperature in Kelvin. Default 273
+        target_pressure (Optional[float]): target pressure in Pa. Default 10**5
+    Returns:
+        MsgGenerator: Plan
+    Yields:
+        Iterator[MsgGenerator]: Bluesky messages
+    """
+    yield from move({temperature: target_temperature, pressure: target_pressure})
+    yield from count(detectors, 1, metadata or {})
```

Comment:
Should: replacing this form of docstring with using Annotated/Field arguments so that the values propagate into json schema?

Reply:
Oh, I did not realise they didn't propagate, that's quite bad. I was hoping to make the plans as vanilla-bluesky as we could, so vanilla python function docstrings are definitely desirable...

Reply:
Lemme get a couple of example plans into blueapi and compare the json schemas

Reply:
```python
@attach_data_session_metadata_decorator()
@validate_call(config={"arbitrary_types_allowed": True})
def linkam_plan_1(
    trajectory: LinkamTrajectory,
    linkam: Linkam3 = LINKAM,
    shutter_time: float = 0.04,
) -> MsgGenerator:
    """
    Follow a trajectory in temperature, collecting a number of frames either at equally
    spaced positions or while continually scanning. e.g. for 2 segments, the first
    stepped and the 2nd flown:\n
    trajectory start   v             v final segment stop\n
                       \\           /\n
       stepped segment__\\__       /\n
                           \\     /  flown segment\n
           1st segment stop \\__ /\n
        exposures:    xx  xx  xx   1/N seconds

    Args:
        trajectory: Trajectory for the scan to follow.
        linkam: Temperature controller
        shutter_time: Time allowed (s) for opening shutter before triggering detectors.

    Returns:
            MsgGenerator: Plan

    Yields:
            Iterator[MsgGenerator]: Bluesky messages
    """
    ...

@attach_data_session_metadata_decorator()
@validate_call(config={"arbitrary_types_allowed": True})
def linkam_plan_2(
    trajectory: Annotated[
        LinkamTrajectory, Field(description="Trajectory for the scan to follow.")
    ],
    linkam: Annotated[Linkam3, Field(LINKAM, description="Temperature controller.")],
    shutter_time: Annotated[
        float,
        Field(
            0.04,
            description="Time allowed for opening shutter before triggering detectors.",
            json_schema_extra={"units": "s"},
        ),
    ],
) -> MsgGenerator:
    """
    Follow a trajectory in temperature, collecting a number of frames either at equally
    spaced positions or while continually scanning. e.g. for 2 segments, the first
    stepped and the 2nd flown:\n
    trajectory start   v             v final segment stop\n
                       \\           /\n
       stepped segment__\\__       /\n
                           \\     /  flown segment\n
           1st segment stop \\__ /\n
        exposures:    xx  xx  xx   1/N seconds
    """
    ...

@attach_data_session_metadata_decorator()
@validate_call(config={"arbitrary_types_allowed": True})
def linkam_plan_3(
    trajectory: LinkamTrajectory = Field(
        description="Trajectory for the scan to follow."
    ),
    linkam: Linkam3 = Field(LINKAM, description="Temperature controller."),
    shutter_time: float = Field(
        0.04,
        description="Time allowed for opening shutter before triggering detectors.",
        json_schema_extra={"units": "s"},
    ),
) -> MsgGenerator:
    """
    Follow a trajectory in temperature, collecting a number of frames either at equally
    spaced positions or while continually scanning. e.g. for 2 segments, the first
    stepped and the 2nd flown:\n
    trajectory start   v             v final segment stop\n
                       \\           /\n
       stepped segment__\\__       /\n
                           \\     /  flown segment\n
           1st segment stop \\__ /\n
        exposures:    xx  xx  xx   1/N seconds
    """
    ...
```

[linkam_plan_1.json](https://github.com/user-attachments/files/17436590/linkam_plan_1.json)
[linkam_plan_2.json](https://github.com/user-attachments/files/17436591/linkam_plan_2.json)
[linkam_plan_3.json](https://github.com/user-attachments/files/17436656/linkam_plan_3.json)

I've just gone through all that just to prove they're basically identical and the Pydantic model doesn't actually get the information about the Fields :upside_down_face:

Reply:
I think we can continue this discussion in #644